### ISSUE TITLE
[Suggestion] Support conditionally setting keys with the spread operator

### ISSUE THREAD

**Issue by ssynix** [Microsoft/TypeScript#16326]
### Background
Support the behavior explained here: https://github.com/tc39/proposal-object-rest-spread/issues/45

This can be supported by extending `...undefined`, `...null` to also ignore `...boolean`.

### Example
There's already an example in the link above.
Personally, this is the snippet of code that I was trying to write:

```ts
this.headers = {
    ...this.headers,
    ...headers,
    ...authToken && { Authorization: authToken },
}
```

### Workaround
```ts
...(authToken ? { Authorization: authToken } : {})
```

**Comment by mhegazy** [comment]
Seems [working for me](http://www.typescriptlang.org/play/index.html#src=var%20authToken%3A%20boolean%20%3D%20%3Cboolean%3Efalse%3B%0D%0A%0D%0A%0D%0Avar%20h%20%3D%20%7B%0D%0A%20%20%20%20...authToken%20%26%26%20%7B%20Authorization%3A%20authToken%20%7D%0D%0A%7D%3B%0D%0A%0D%0Aconsole.log(JSON.stringify(h))%3B)

**Comment by ssynix** [comment]
Try it with `strictNullChecks`.

**Comment by sandersn** [comment]
A fix is up at #18451, but it's kind of clunky &mdash; I had to special-case `false | T` to produce [the equivalent of] `Partial<T>`.
